RFC-0005-multi-cluster-resource-management.md

```diff
@@ -0,0 +1,64 @@
+3. Load-balancing queries effectively across multiple clusters requires the load balancer to understand the resource group and queue state of each cluster. This is difficult to achieve, and in practice, stale information leads to poor load balancing. 
+4. Since the query is queued on the cluster itself, restarting clusters for maintenance reasons requires either that the queued queries be killed, or a lengthy drain time. 
+
+## Proposed Implementation
+
+We propose to add support in Presto for an optional queueing service that sits upstream of multiple Presto clusters. This queueing service will serve as the global entry-point for all queries from the client and will implement the QueuedStatement protocol. This queueing service is responsible for: 
+
+1. Authenticating the request 
+2. Adding the query to the global queue 
+3. Responding to clients polling for the queueing status of their query
+4. Picking the next query to execute 
+5. Selecting the best cluster for execution of this query
+6. Dispatching this query to the cluster
+7. Redirecting the client to the picked cluster
+
+The internal details of this queueing service are still under development, and this document focuses on the changes required to Presto for such a service to work. If there is interest from the community, we can provide an open-source reference implementation of this queueing service, once mature. 
+
+### Diagram
+![Presto Github Issue GRM](https://github.com/user-attachments/assets/7b2af8fd-5d82-411d-ac97-795e54e35164)
+
+### Work Items
+Because we want this to be transparent to Presto clients, the front-end of the queueing service will reuse the existing functionality of classes in the presto-dispatcher package such as QueuedStatementResource, DispatchManager, DispatchQuery, etc. The high-level work items are:
+
+1. Allow for the queueing service to forward all required query metadata to the cluster, such as query ID, authenticated identity, query state timings, etc. This will involve changes to QueuedStatementResource, QueryStateMachine, and DispatchManager, so that coordinator can create a query with an already populated QueryId and AuthorizedIdentity.
+2. De-couple the dispatching and execution phases of the query lifecycle. This will involve separating and removing dependencies that are required for execution of the query, but not for queueing, such as Metadata class. We will create a new module, presto-dispatcher, that has all classes required for the queueing phase, such as QueuedStatementResource, DispatchManager, ResourceGroupManager, etc. The presto-main module will have a dependency on this new module. Interfaces such as ResourceGroupManager and DispatchQuery will be in presto-dispatcher, with implementations such as InternalResourceGroupManager and LocalDispatchQuery in presto-main. In addition, it could be necessary to move some classes into a common module, if they are required by both presto-main and presto-dispatcher.
```

Comment:
> We will create a new module, presto-dispatcher
Will presto-dispatcher be used by the new service?

Reply:
Yes, the queueing service will have a dependency on presto-dispatcher. The queueing service "implementation" of the dispatcher is as standalone server, and the presto-main "implementation" of the dispatcher will be in-process with the coordinator. The code is already organized this way - for example, DispatchQuery is an interface that would live in presto-dispatcher, with LocalDispatchQuery being the implementation in presto-main and a hypothetical RemoteDispatchQuery that could be implemented by the queueing service.